docs/machine-learning/ml-profiling.md

```diff
@@ -0,0 +1,120 @@
+and sorting parameters for better analysis experience.
+
+Below you can find an example of using *cProfile* to profile a chunk of code.
+
+```python
+import cProfile
+
+# Start profiling
+profiler = cProfile.Profile()
+profiler.enable()
+
+# -- YOUR CODE GOES HERE ---
+
+# Stop profiling
+profiler.disable()
+
+# Write profiler results to an html file
+profiler.dump_stats("profiler_results.prof")
+```
+
+> Note: one epoch of model training is usually enough for profiling. There's no need to run more epochs and produce
+additional cost.
+
+More information on *cProfiler*:
+
+- [The Python Profilers](https://docs.python.org/3/library/profile.html)
+
+## PyTorch model training profiling
+
+PyTorch 1.8 includes an updated PyTorch
+[profiler](https://pytorch.org/blog/introducing-pytorch-profiler-the-new-and-improved-performance-tool/)
+that is supplied together with the PyTorch distribution and doesn't require any additional installation.
+Using *PyTorch profiler* one can record CPU side operations as well as CUDA kernel launches on GPU side.
+The profiler can visualize analysis results using TensorBoard plugin as well as provide suggestions
+on bottlenecks and potential code improvements.
+
+```python
+ with torch.profiler.profile(
+    # Limit number of training steps included in profiling
+    schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=2),
+    # Automatically saves profiling results to disk
+    on_trace_ready=torch.profiler.tensorboard_trace_handler,
+    with_stack=True
+) as profiler:
+    for step, data in enumerate(trainloader, 0):
+        # -- TRAINING STEP CODE GOES HERE ---
+        profiler.step()
+```
+
+The tensorboard_trace_handler for analysis in TensorBoard. Those can be visualized by installing TensorBoard
+plugin and running TensorBoard on your log directory.
+
+```bash
+pip install torch_tb_profiler
+tensorboard --logdir=<LOG_DIR_PATH>
+# Navigate to `http://localhost:6006/#pytorch_profiler`
+```
+
+> Note: make sure to provide right parameters to the `torch.profiler.schedule`. Usually you need several steps of training to be profiled rather than the whole epoch.
+
+More information on *PyTorch profiler*:
+
+- [PyTorch Profiler Recipe](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html)
+- [Introducing PyTorch Profiler - the new and improved performance tool](https://pytorch.org/blog/introducing-pytorch-profiler-the-new-and-improved-performance-tool/)
+
+## Azure Machine Learning pipeline profiling
+
+In our projects we often use [Azure Machine Learning](https://azure.microsoft.com/en-us/services/machine-learning/)
+pipelines to train Machine Learning models. Most of the profilers can also be used in conjunction with Azure Machine Learning.
+In a nutshell, you need to choose a profiler, that meets the following criteria:
+
+- The profiler can be enabled and disabled inside of the code
```

Comment:
This isn't very clear. Do you mean that one can control whether the profiler is on or off using a parameter, or that certain parts of the code would be profiled and others won't? Consider rephrasing.

Reply:
Thanks! Great suggestion